File: js/src/providers/tavily.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

const MODE: Record<string, string> = { fast: "fast", balanced: "basic", deep: "advanced" };

export const tavily: ProviderSpec = {
  name: "tavily",
  envKeys: ["TAVILY_API_KEY"],
  defaultBaseUrl: "https://api.tavily.com",
  extraPackage: "@tavily/core",
  nativeImport: { module: "@tavily/core", exports: ["tavily", "TavilyClient"] },
  capabilities: new Set(["domains", "country", "date", "mode", "answer", "content", "news"]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const body: Record<string, unknown> = {
      query: req.query,
      max_results: Math.min(req.maxResults, 20),
      search_depth: MODE[req.mode ?? ""] ?? "basic",
      topic: req.searchType === "news" ? "news" : "general",
    };
    if (req.includeDomains.length) body.include_domains = req.includeDomains;
    if (req.excludeDomains.length) body.exclude_domains = req.excludeDomains;
    if (req.startPublishedDate) body.start_date = req.startPublishedDate.slice(0, 10);
    if (req.endPublishedDate) body.end_date = req.endPublishedDate.slice(0, 10);
    if (req.country) body.country = req.country.toLowerCase();
    if (req.answer) body.include_answer = true;
    if (req.includeContent) body.include_raw_content = true;
    Object.assign(body, req.extra);
    return {
      method: "POST",
      url: `${ctx.baseUrl}/search`,
      headers: { Authorization: `Bearer ${ctx.apiKey ?? ""}`, "Content-Type": "application/json" },
      json: body,
    };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    const results = (data.results ?? []).map((item: any) =>
      result({
        title: item.title,
        url: item.url,
        snippet: item.content,
        text: item.raw_content,
        highlights: item.content ? [item.content] : [],
        score: item.score,
        publishedDate: item.published_date,
        raw: item,
      }),
    );
    return finalize(tavily, req, results, data, {
      answer: data.answer,
      requestId: data.request_id,
      elapsedMs,
    });
  },
};

File: python/src/anysearch/providers/tavily.py
"""Tavily — search API purpose-built for LLMs, with answers and raw content."""

from __future__ import annotations

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability

_MODE = {"fast": "fast", "balanced": "basic", "deep": "advanced"}


class TavilyProvider(BaseProvider):
    name = "tavily"
    env_keys = ("TAVILY_API_KEY",)
    default_base_url = "https://api.tavily.com"
    extra_package = "tavily-python"
    native_client = ("tavily", ("TavilyClient",))
    capabilities = frozenset(
        {
            Capability.DOMAINS,
            Capability.COUNTRY,
            Capability.DATE,
            Capability.MODE,
            Capability.ANSWER,
            Capability.CONTENT,
            Capability.NEWS,
        }
    )

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        body = {
            "query": req.query,
            "max_results": min(req.max_results, 20),
            "search_depth": _MODE.get(req.mode, "basic"),
            "topic": "news" if req.search_type == "news" else "general",
        }
        if req.include_domains:
            body["include_domains"] = req.include_domains
        if req.exclude_domains:
            body["exclude_domains"] = req.exclude_domains
        if req.start_published_date:
            body["start_date"] = req.start_published_date[:10]
        if req.end_published_date:
            body["end_date"] = req.end_published_date[:10]
        if req.country:
            body["country"] = req.country.lower()
        if req.answer:
            body["include_answer"] = True
        if req.include_content:
            body["include_raw_content"] = True
        body.update(req.extra)
        return PreparedRequest(
            method="POST",
            url=f"{self.base_url}/search",
            headers={"Authorization": f"Bearer {self.api_key}", "Content-Type": "application/json"},
            json=body,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        results = []
        for item in data.get("results", []) or []:
            content = item.get("content")
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("url"),
                    snippet=content,
                    text=item.get("raw_content"),
                    highlights=[content] if content else [],
                    score=item.get("score"),
                    published_date=item.get("published_date"),
                    raw=item,
                )
            )
        return self._response(
            req,
            results,
            data,
            answer=data.get("answer"),
            request_id=data.get("request_id"),
            elapsed_ms=elapsed_ms,
        )
